examples/structured_data/classification_with_tfdf.py

```diff
@@ -0,0 +1,726 @@
+    return feature_usages
+
+
+"""
+### Create a Gradient Boosted Trees model
+
+When compiling a decision forests model, you may only provide extra evaluation metrics.
+The loss is specified in the model construction,
+and the optimizer is irrelevant to decision forests models.
+"""
+
+
+def create_gbt_model():
+    feature_usages = specify_feature_usages(create_model_inputs())
+    gbt_model = tfdf.keras.GradientBoostedTreesModel(
+        features=feature_usages,
+        exclude_non_specified_features=True,
+        growing_strategy=GROWING_STRATEGY,
+        num_trees=NUM_TREES,
+        max_depth=MAX_DEPTH,
+        min_examples=MIN_EXAMPLES,
+        subsample=SUBSAMPLE,
+        validation_ratio=VALIDATION_RATIO,
+        task=tfdf.keras.Task.CLASSIFICATION,
+        loss="DEFAULT",
+    )
+
+    gbt_model.compile(metrics=[keras.metrics.BinaryAccuracy(name="accuracy")])
+    return gbt_model
+
+
+"""
+### Train and evaluate the model
+"""
+
+gbt_model = create_gbt_model()
+history = run_experiment(gbt_model, train_data, test_data)
+
+"""
+### Inspect the model
+
+The `model.summary()` method will display several types of information about
+your decision trees model, model type, task, input features, and feature importance.
+"""
+
+print(gbt_model.summary())
+
+"""
+## Experiment 2: Decision Forests with target encoding
+
+[Target encoding](https://dl.acm.org/doi/10.1145/507533.507538) is a common preprocessing
+technique for categorical features that convert them into numerical features.
+Using categorical features with high cardinality as-is may lead to overfitting.
+Target encoding aims to replace each categorical feature value with one or more
+numerical values that repesent its co-occerence with the target labels.
+
+More preciely, given a categorical feature, the binary target encoder in this example
+will produce three new numerical features:
+
+1. `positive_frequency`: How many times each feature value occured with a positive target label.
+2. `negative_frequency`: How many times each feature value occured with a negative target label.
+3. `positive_probability`: The probability that the target label is possitive,
+given the feature value, which is computed as
+`positive_frequency / (positive_frequency + negative_frequency)`.
+
+
+Note that target encoding is effective with models that cannot automatically
+learn dense repesentations to categorical features, such as decision forests
+or kernel methods. If neural network models are used, its recommended to
+encode categorical features as embeddings.
+"""
+
+"""
+### Implement Binary Target Encoder
+"""
+
+from keras.engine import base_preprocessing_layer
+
+
+class BinaryTargetEncoding(base_preprocessing_layer.PreprocessingLayer):
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        base_preprocessing_layer.keras_kpl_gauge.get_cell("BinaryTargetEncoding").set(
+            True
+        )
+
+    def adapt(self, data):
```

Comment:
With pandas, Target mean can be done in two lines:

```python
ds = pd.DataFrame(
    {"cat_feature":["a","b","c","a","b","c"],
     "binary_label":[0,0,1,0,1,1]})

target_map = ds.groupby("cat_feature")["binary_label"].mean()
ds["taget_mean_feature"] = ds["cat_feature"].map(target_map)
```

Maybe something similar could be done here. Note that group_by is [available](https://www.tensorflow.org/api_docs/python/tf/data/experimental/group_by_reducer) in tf.